proofs/eo/cpc/rules/Arith.eo

```diff
@@ -343,44 +349,47 @@
+  (Bool Bool) Bool
   (
-    (($mk_arith_mult_abs_comparison_lhs = (and (= (abs t) (abs u)) B))  (eo::cons * t ($mk_arith_mult_abs_comparison_lhs = B)))
-    (($mk_arith_mult_abs_comparison_lhs > (and (> (abs t) (abs u)) B))  (eo::cons * t ($mk_arith_mult_abs_comparison_lhs > B)))
-    (($mk_arith_mult_abs_comparison_lhs > (and (and (= (abs t) (abs u)) 
-                                                    (not (= t z))) B))  (eo::requires (eo::to_z z) 0 
-                                                                        (eo::cons * t ($mk_arith_mult_abs_comparison_lhs > B))))
-    (($mk_arith_mult_abs_comparison_lhs r true)                            1)
+    (($mk_arith_mult_abs_comparison_rec (and (r (abs t) (abs u)) B) (r a b))
+      ($mk_arith_mult_abs_comparison_rec B
+        (r (eo::list_concat * a (* t)) (eo::list_concat * b (* u)))))
+    (($mk_arith_mult_abs_comparison_rec (and (and (= (abs t) (abs u))
+                                                  (not (= t z))) B) (> a b))
+      (eo::requires (eo::to_z z) 0
+        ($mk_arith_mult_abs_comparison_rec B
+          (> (eo::list_concat * a (* t)) (eo::list_concat * b (* u))))))
+    (($mk_arith_mult_abs_comparison_rec true (r a b))  (r (abs a) (abs b)))
   )
 )
 
-; program: $mk_arith_mult_abs_comparison_rhs
+; program: $mk_arith_mult_abs_comparison
 ; args:
-; - r (-> T U Bool): The relation.
-; - F Bool: A conjunction of the remaining premises we have left to process.
+; - F Bool: A conjunction of the remaining premises.
 ; return: >
-;   The right hand side of the equality proven by arith_mult_abs_comparison for
-;   r and the remaining premises F.
-; note: This program is analogous to $mk_arith_mult_abs_comparison_lhs.
-(program $mk_arith_mult_abs_comparison_rhs ((T Type) (U Type) (r (-> T U Bool)) (t T) (u U) (z T) (B Bool :list) (S Type))
-  ((-> T U Bool) Bool) S
+;   The literal proven by arith_mult_abs_comparison given premises F.
+(program $mk_arith_mult_abs_comparison
+  ((T Type) (U Type) (V Type) (r T) (t T) (u U) (z V) (B Bool :list) (S Type))
+  (Bool) Bool
   (
-    (($mk_arith_mult_abs_comparison_rhs = (and (= (abs t) (abs u)) B))  (eo::cons * u ($mk_arith_mult_abs_comparison_rhs = B)))
-    (($mk_arith_mult_abs_comparison_rhs > (and (> (abs t) (abs u)) B))  (eo::cons * u ($mk_arith_mult_abs_comparison_rhs > B)))
-    (($mk_arith_mult_abs_comparison_rhs > (and (and (= (abs t) (abs u))
-                                                    (not (= t z))) B))  (eo::requires (eo::to_z z) 0
-                                                                        (eo::cons * u ($mk_arith_mult_abs_comparison_rhs > B))))
-    (($mk_arith_mult_abs_comparison_rhs r true)                           1)
+    (($mk_arith_mult_abs_comparison (and (> t u) B))
+      ($mk_arith_mult_abs_comparison_rec (and (> t u) B) (> 1 1)))
+    (($mk_arith_mult_abs_comparison B)
+      ($mk_arith_mult_abs_comparison_rec B (= 1 1)))
```

Comment:
`(> 1 1)` is not true... How about the following:
```suggestion
    (($mk_arith_mult_abs_comparison (and (> t u) B))
      ($mk_arith_mult_abs_comparison_rec B (> (* t) (* u))))
    (($mk_arith_mult_abs_comparison (and (= t u) B))
      ($mk_arith_mult_abs_comparison_rec B (= (* t) (* u))))
```

Reply:
Yes, I like this.